### PR TITLE
 refactor: use async function and support koa@2

### LINKED ISSUE #10: Koa 2
If this library is still being maintained it would be nice to have a `next` tag for `koa@2`. Especially since koa-roles has its own `.use` methods which yield.

### PR DESCRIPTION
**Pull request by dead-horse**
closes #10

### CODE REVIEW

**Comment by dead-horse** [comment]
@fengmk2

File: README.md [hunk 1]
```diff
@@ -34,22 +34,22 @@ $ npm install koa-roles
 ```js
 var Roles = require('koa-roles');
 var koa = require('koa');
```
Comment (fengmk2) [inline README.md]: Koa and new Koa()

File: README.md [hunk 2]
```diff
@@ -34,22 +34,22 @@ $ npm install koa-roles
...
-var app = koa();
+var app = new koa();
 
 var user = new Roles({
-  failureHandler: function *(action) {
+  async failureHandler(action, ctx) {
```
Comment (fengmk2) [inline README.md]: the first argument should be ctx.

```js
async failureHandler(ctx, action) {
```

File: index.js [hunk 3]
```diff
@@ -147,17 +129,14 @@ KoaRoles.prototype.use2 = function (action, fn) {
...
-      yield roles.failureHandler.call(this, action);
-    } else {
-      roles.failureHandler.call(this, action);
+KoaRoles.prototype.can = function(action) {
+  const roles = this;
+  return async (ctx, next) => {
+    if (await roles.test(ctx, action)) {
+      return next();
     }
+    const r = roles.failureHandler.call(ctx, action, ctx);
```
Comment (fengmk2) [inline index.js]: `roles.failureHandler(ctx, action)` should breaking change.

**Comment by dead-horse** [comment]
重构了

**Comment by fengmk2** [comment]
2.0.0

**Pull request merged**